### PR TITLE
Integrating Collision Detector 2 ©

### PR DESCRIPTION
**Pull request by Peter-JanGootzen**
# Description
This PR integrates the new Collision Detector 2 ©, check [the PR on BrickEngine](https://github.com/Brick-Studios/BrickEngine/pull/58) for more info on that.
This PR also contains:
- Fixed bullets not despawning when the last bullet kills a player. This is pretty much dubble functionality with DespawnComponent, however there is no other way to fix this issue then by optionally despawning a bullet instantly when dealing damage.
- Fixed the WeaponComponent not containing the collider that bullet is going to get, we used to hard coded create a new Collider that was just as big as the Transform of the bullet.

# How can this be tested?

Try to collide with as many things in as many different ways as possible. And focus on hitting things in the corners, because if the velocity is fast and the direction in a slighty off tilted up angle, some weird displacing of the entities could happen (we have not been able to spot this).

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] Code is const correct

WhatsApp Image 2019-11-30 at 22 58 20
WhatsApp Image 2019-11-30 at 23 22 14
Capture

### CODE REVIEW

**Review by LuukSantegoeds — DISMISSED** [review]
As said in the Engine: "Clearly the artwork is way better than the code" and this PR is no exception.

_This is the spot where I would show you a snippet of the build failing but the error was too long :P_

File: lib/controllers/game_controller.cpp [hunk 1]
```diff
@@ -71,7 +72,14 @@ GameController::GameController() {
     engine->start();
     entityManager = std::make_shared<EntityManager>();
     entityFactory = std::make_shared<EntityFactory>(entityManager, *engine->getRenderableFactory());
-    collisionDetector = std::make_shared<CollisionDetector>(entityManager);
+
+    std::unordered_map<std::string, std::set<std::string>> is_trigger_exceptions;
+    is_trigger_exceptions.insert({ "Critter", std::set<std::string> { "Platform" } });
+    is_trigger_exceptions.insert({ "Weapon", std::set<std::string> { "Platform" } });
+    // For when the players are ded
+    is_trigger_exceptions.insert({ "Player", std::set<std::string> { "Platform" } });
```
Comment (MarkMeer) [inline lib/controllers/game_controller.cpp]: What does this do? Maybe add a comment about its function as 'is_trigger_exceptions' looks like it has something to do with an boolean to show whether it a trigger exception. However, it has a string and a set with strings.

**Review by LuukSantegoeds — CHANGES_REQUESTED** [review]
Change the in-air recoil and I approve

**Review by MarkMeer — APPROVED** [review]
Looking good, functionally a few 🐛 but we've discussed those. 


edit: Meant to comment on this, not approve.

File: lib/controllers/game_controller.cpp (hunk 1 above)
Comment (JanSchollaert) [reply lib/controllers/game_controller.cpp]: Those are for triggers, all triggers have no collision and will fall through everything, critter and weapons should not fall through platforms, so we made trigger exceptions using tags. The entities with the left tag will now collide with entities with the right tag.

Comment (JanSchollaert) [reply lib/controllers/game_controller.cpp]: I put a comment in the header where it's made

**Review by LuukSantegoeds — APPROVED** [review]
Works gr8

**Review by MarkMeer — APPROVED** [review]
Plays nice 👍

**Pull request merged**